### PR TITLE
Jamie/4439 - Bump Angular from 10 to 11

### PR DESCRIPTION
**Pull request by SEAjamieD**
### :nut_and_bolt: Description: What code changed, and why?

This PR bumps our Angular framework from v10 to v11, which is the most current production ready version of Angular. 

#### Migrate Angular

This did two things:

1. Update the various `@angular` packages in package.json.
2. Automatically apply migration patches to our code base.

Commands:
`ng update @angular/core @angular/cli`
`ng update @angular/material`

Reference: https://update.angular.io/?l=3&v=10.2-11.0

#### Update all other npm packages to most recent compatible version.
npm outdated will list the following constrained packages.  This was done with a combination
of `npm outdated`,  `ng update <package>` and `npm audit` - more details provided in README.md
```
Package                Current    Wanted       Latest  Location
@ngrx/effects          10.1.0  10.1.0       10.1.2  automate-ui
@ngrx/entity           10.1.0  10.1.0       10.1.2  automate-ui
@ngrx/router-store     10.1.0  10.1.0       10.1.2  automate-ui
@ngrx/store            10.1.0  10.1.0       10.1.2  automate-ui
@ngrx/store-devtools   10.1.0  10.1.0       10.1.2  automate-ui
immutable               3.8.2   3.8.2  4.0.0-rc.12  automate-ui
jwt-decode              2.2.0   2.2.0        3.1.2  automate-ui
karma                   5.1.1   5.1.1        5.2.3  automate-ui
rrule                   2.4.1   2.4.1        2.6.6  automate-ui
typescript              4.0.5   4.0.5        4.1.3  automate-ui
zone.js                0.10.3  0.10.3       0.11.3  automate-ui
```

#### Patch code incompatibilities due to all of the above

The few necessary patches, included in various commits herein, have achieved these milestones:

 - [x] build (`ng build`)
 - [x] unit tests (via wallaby in VSCode)
 - [x] unit tests (`make unit` via command-line)
 - [x] e2e tests (protractor via `make e2e`)
 - [x] lint (`make lint`)
 - [x] launch application (`make serve`)
 - [x] application runs with no errors in browser
 - [x] application runs with no warnings in browser

### :chains: Related Resources
Last Angular Update (v9 -> v10): https://github.com/chef/automate/pull/4468

### :+1: Definition of Done

With Angular 11 / TypeScript 4 ecosystem, Automate-UI and all of its unit and integration tests still function properly.
After running `npm install`, 0 vulnerabilities should be found.

### :athletic_shoe: How to Build and Test the Change

Do your usual hab studio setup notably including building `automate-ui-devproxy`. 

Then (line numbers shown for reference):
```
[1] cd components/automate-ui
[2] mv node_modules /location/to/save/your/old/node_modules
[3] npm install
[4] make serve
```

Notes:
Line [2] is only needed if you want to be able to easily resume what you were doing before you tried out this PR! Once you are done evaluating, move or delete your new node_modules and restore your copy made in line [2].
Line [3] generates the new node_modules directory for this PR.
Line [4] starts automate-ui.

### :white_check_mark: Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] Tests added/updated?
- [x] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by SEAjamieD** [comment]
Note: ngrx 10.1.1 throws a missing peer dependency of angular 10, I believe this can be safely disregarded until ngrx 11 is released.  Reference: https://github.com/ngrx/platform/issues/2830\#issuecomment-748628514

**Comment by SEAjamieD** [comment]
Note: Formbuilder#group deprecations: https://angular.io/api/forms/FormBuilder#group

File: components/automate-ui/README.md [hunk 1]
```diff
@@ -55,24 +55,17 @@ Certain packages in package.json are constrained for the reasons detailed here.
...
-  ../node_modules/@types/jasmine/ts3.1/index.d.ts:22:1
-    22 type ImplementationCallback = jasmine.ImplementationCallback;
-       ~~~~
-    Conflicts are in this file.
-```
+### Package jwt-decode: ^2.2.0
 
-There are very few reports of others experiencing this problem and no resolution presents itself.
-Reference: https://stackoverflow.com/a/57592510
+Newer versions of jwt_decode introduce a new import format, which is breaking some of our testing.
```
Comment (SEAjamieD) [inline components/automate-ui/README.md:68]: We should spend some time resolving this so that we can upgrade to the newest version of jwt_decode, though I think that can be taken care of in its own PR.

File: components/automate-ui/README.md [hunk 2]
```diff
@@ -88,6 +81,9 @@ Reason: Later releases are release candidates; should only be using production-r
 
 Per https://github.com/chef/automate/pull/1867, future versions have made a breaking change so it breaks our code base.
 
+### Package zone.js: ^0.10.3
+
+Per: https://www.npmjs.com/package/zone.js?activeTab=readme, Starting with v0.11.1, A breaking change is introduced for legacy browsers such as IE11.  If we are no longer supporting IE 11, we are safe to upgrade, otherwise we can instead `import 'zone.js/dist/zone';`
```
Comment (SEAjamieD) [inline components/automate-ui/README.md:86]: If we're comfortable with not supporting IE 11, we can update.  If not we can use a backwards compatible import

File: components/automate-ui/package.json [hunk 3]
```diff
@@ -7,6 +7,7 @@
     "start": "npm run serve",
     "serve": "ng serve --configuration=no_auth",
     "serve:hab": "echo '*** Please ignore the message to go to http://localhost:4200 -- the correct URL is https://a2-dev.test/ ***' && ng serve --host 0.0.0.0 --disable-host-check --public-host=httpfake://0.0.0.0:0/ --ssl-key ../../dev/certs/automate-ui.key --ssl-cert ../../dev/certs/automate-ui.crt --base-href / --live-reload false --poll 1000",
+    "preinstall": "npx npm-force-resolutions",
```
Comment (SEAjamieD) [inline components/automate-ui/package.json:10]: This was necessary in order to fix the security vulnerability coming from the transitive dependency, `minimist`

File: components/automate-ui/package.json [hunk 4]
```diff
@@ -65,42 +68,45 @@
...
-    "tslint": "^5.0.0",
+    "ts-node": "^9.1.1",
+    "tslint": "^6.1.3",
     "tslint-defocus": "^2.0.6",
     "typescript": "~4.0.5",
     "wallaby-webpack": "^3.9.16"
+  },
+  "resolutions": {
+    "minimist": "^1.2.5"
   }
```
Comment (SEAjamieD) [inline components/automate-ui/package.json:111]: This forces any packages depending on `minimist` to use `^1.2.5` - which help solve our `minimist` security vulnerability

File: components/automate-ui/src/app/pages/job-add/job-add.component.ts [hunk 5]
```diff
@@ -93,14 +93,14 @@ export class JobAddComponent implements OnDestroy {
     const nodesGroup = this.fb.group({
       managers: this.fb.array([])
     }, {
-      validator: nodeSelectionRequiredValidator
+      validators: nodeSelectionRequiredValidator
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/pages/job-add/job-add.component.ts:96]: validator was deprecated as part of the FormBuilder#group deprecation in favor of Abstract control and `validators`
ref: https://angular.io/api/forms/FormBuilder#group

File: components/automate-ui/src/app/pages/job-add/job-add.validators.ts [hunk 6]
```diff
@@ -1,19 +1,19 @@
...
+  AbstractControl,
   FormArray,
-  FormGroup,
   ValidationErrors,
   ValidatorFn
 } from '@angular/forms';
 
 export const nodeSelectionRequiredValidator: ValidatorFn =
-  (control: FormGroup): ValidationErrors | null => {
+  (control: AbstractControl): ValidationErrors | null => {
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/pages/job-add/job-add.validators.ts:9]: Abstract control is the replacement for `FormGroup` in this case as a result of the FormBuilder#group deprecation.
ref: https://angular.io/api/forms/FormBuilder#group

File: components/automate-ui/src/app/pages/node-details/node-details.component.spec.ts [hunk 7]
```diff
@@ -40,7 +40,7 @@ describe('NodeDetailsComponent', () => {
     describe('on initialization', () => {
       it('subscribes to the node-details event service', () => {
         spyOn(eventService.showModal$, 'subscribe')
-          .and.returnValue(new BehaviorSubject<boolean>(false));
+          .and.returnValue(new Subscription);
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/pages/node-details/node-details.component.spec.ts:43]: The upgrade picked up that the return on this was actually a Subscription and needed to be updated as such.

**Comment by msorens** [comment]
The link above for  https://github.com/ngrx/platform/issues/2830\#issuecomment-748628514 is broken.

File: components/automate-ui/README.md (hunk 2 above)
Comment (msorens) [reply components/automate-ui/README.md:86]: Since Microsoft is just starting de-commissioning support for IE 11--with some products still supporting it throughout 2021--I am thinking we should probably be conservative and continue supporting it, too.

File: components/automate-ui/src/app/pages/job-add/job-add.validators.ts (hunk 6 above)
Comment (msorens) [reply components/automate-ui/src/app/pages/job-add/job-add.validators.ts:9]: I am not seeing any mention of `FormGroup` being deprecated. 😕

**Comment by msorens** [comment]
Welcome, @SEAjamieD. 

Many commits to this PR -- that cleanly and clearly documents your steps. ✅ 

You have completed level `N`. Time to advance to level `N+1`.
image

Observe the ellipses on many of the commit messages (image below). That is a polite (yet subtle 😉 )way for the system to tell you your message _headers_ (as distinct from message _bodies_) are too long. Have a look at [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/) for ways to refine the quality of your commit messages.

image

File: components/automate-ui/README.md [hunk 8]
```diff
@@ -55,24 +55,17 @@ Certain packages in package.json are constrained for the reasons detailed here.
...
-       ~~~~
-    Conflicts are in this file.
-```
+### Package jwt-decode: ^2.2.0
 
-There are very few reports of others experiencing this problem and no resolution presents itself.
-Reference: https://stackoverflow.com/a/57592510
+Newer versions of jwt_decode introduce a new import format, which is breaking some of our testing.
 
 ### Package tslint: ^5.0.0
```
Comment (msorens) [inline components/automate-ui/README.md:70]: tslint seems to be at latest so OK to delete this `tslint` entry here.

File: components/automate-ui/README.md [hunk 9]
```diff
@@ -55,24 +55,17 @@ Certain packages in package.json are constrained for the reasons detailed here.
 At any future moment though, the reasons for constraint here could be invalidated, so this should be updated as needed when package.json is updated.
 While it is problematic to document this information due to the maintenance burden, the value of having this in one place outweighs that burden.
```
Comment (msorens) [inline components/automate-ui/README.md:57]: I suspect this package just bumped from 165 to 166 in the midst of your work and could be updated to remove it from our exceptions list.
```
@types/lodash         4.14.165  4.14.166     4.14.166
```

Comment (msorens) [inline components/automate-ui/README.md:57]: Notice that `npm outdated` is showing these:
```
Package                Current    Wanted       Latest  Location
@ngrx/effects           10.1.0    10.1.2       10.1.2  automate-ui
@ngrx/entity            10.1.0    10.1.2       10.1.2  automate-ui
@ngrx/router-store      10.1.0    10.1.2       10.1.2  automate-ui
@ngrx/store             10.1.0    10.1.2       10.1.2  automate-ui
@ngrx/store-devtools    10.1.0    10.1.2       10.1.2  automate-ui
```
If you really want to constrain these, as you have mentioned, so it does not increment even the last segment (the patch number), you need to update package.json to use exact version (e.g. `10.1.0` instead of `^10.1.0`). Otherwise, IIRC, if you just run `npm update` right now, it will bump up to the "wanted" version. (You could try to see if that is true.).
Also, consider adding an entry for these (e.g. `@ngrx/*`) in the readme.md file.

**Review by msorens — COMMENTED** [review]
Very nice job, and well-documented!

File: components/automate-ui/src/app/pages/job-add/job-add.validators.ts (hunk 6 above)
Comment (SEAjamieD) [reply components/automate-ui/src/app/pages/job-add/job-add.validators.ts:9]: From about halfway down on https://angular.io/api/forms/FormBuilder#group

>Construct a new FormGroup instance.
>
>~~`group(controlsConfig: { [key: string]: any; }, options: { [key: string]: any; }): FormGroup`~~
>
>**Deprecated** This API is not typesafe and can result in issues with Closure Compiler renaming. Use the FormBuilder#group overload with AbstractControlOptions instead. Note that AbstractControlOptions expects validators and asyncValidators to be valid validators. If you have custom validators, make sure their validation function parameter is AbstractControl and not a sub-class, such as FormGroup. These functions will be called with an object of type AbstractControl and that cannot be automatically downcast to a subclass, so TypeScript sees this as an error. For example, change the (group: FormGroup) => ValidationErrors|null signature to be (group: AbstractControl) => ValidationErrors|null.

The main portion for our concerns here are these two sentences:
>If you have custom validators, make sure their validation function parameter is AbstractControl and not a sub-class, such as FormGroup. 

>change the (group: FormGroup) => ValidationErrors|null signature to be (group: AbstractControl) => ValidationErrors|null.

Since these are custom validators, they needed to be updated with `AbstractControl`.  I must be mistaken in saying that FormBuilder#group is deprecated - the paragraph above will have to stand in here as a more accurate definition.

File: components/automate-ui/README.md (hunk 2 above)
Comment (SEAjamieD) [reply components/automate-ui/README.md:86]: We'll have to wait to upgrade zone.js to `0.11.1` until the release of `angular/core 11.1.0` otherwise we'll get an incompatible peer dependency warning.  Should be cleared up in the next release re: https://github.com/angular/angular/issues/39094#issuecomment-742227156

File: components/automate-ui/README.md (hunk 9 above)
Comment (SEAjamieD) [reply components/automate-ui/README.md:57]: Thanks for finding this!

Comment (SEAjamieD) [reply components/automate-ui/README.md:57]: I removed the `^`'s and left at `10.1.0` because of peer dependencies in `angular/core` - looks like these should be resolved soon with the release of ngrx 11.  From what I've been able to uncover, it appears we would still be safe to `--force`, but i'd prefer to wait until the official ngrx 11 release.

I've updated in the README as well

**Review by msorens — APPROVED** [review]
